src/aws_credentials.pas

```diff
@@ -75,8 +108,164 @@   TAWSSignatureVersion3 = class sealed(TAWSAbstractSignature)
     function Calculate(Request: IHTTPRequest): string; override;
   end;
 
+  { TAWSSignatureVersion4 }
+
+  TAWSSignatureVersion4 = class sealed(TAWSAbstractSignature)
+  private
+    function BuildHeader(const sHeader: String): String;
+    procedure SignedHeaders(const sHeader: String; var sToSing, sToCanonical: String);
+  public
+    function Calculate(Request: IHTTPRequest): string; override;
+  end;
+
 implementation
 
+{ TAWSSignatureHMAC256 }
+
+constructor TAWSSignatureHMAC256.Create(const sAccessKey, sDataStamp,
+  sRegionName, sServiceName: string);
+begin
+  inherited Create;
+  FAccessKey:= sAccessKey;
+  FDataStamp:= sDataStamp;
+  FRegionName:= sRegionName;
+  FServiceName:= sServiceName;
+end;
+
+class function TAWSSignatureHMAC256.New(const sAccessKey, sDataStamp,
+  sRegionName, sServiceName: string): IAWSSignatureHMAC256;
+begin
+  Result := Create(sAccessKey, sDataStamp, sRegionName, sServiceName);
+end;
+
+function TAWSSignatureHMAC256.AccessKey: string;
+begin
+  Result := FAccessKey;
+end;
+
+function TAWSSignatureHMAC256.DataStamp: string;
+begin
+  Result := FDataStamp;
+end;
+
+function TAWSSignatureHMAC256.RegionName: string;
+begin
+  Result := FRegionName;
+end;
+
+function TAWSSignatureHMAC256.ServiceName: string;
+begin
+  Result := FServiceName;
+end;
+
+function TAWSSignatureHMAC256.Signature: TSHA256Digest;
+var
+  oSHA256: TSHA256Digest;
+begin
+  HMAC_SHA256(UTF8Encode('AWS4'+FAccessKey), UTF8Encode(FDataStamp), oSHA256);
+  HMAC_SHA256(oSHA256, UTF8Encode(FRegionName), oSHA256);
+  HMAC_SHA256(oSHA256, UTF8Encode(FServiceName), oSHA256);
+  HMAC_SHA256(oSHA256, UTF8Encode('aws4_request'), oSHA256);
+  Result := oSHA256;
+end;
+
+{ TAWSSignatureVersion4 }
+
+function TAWSSignatureVersion4.BuildHeader(const sHeader: String): String;
+var
+  i: Integer;
+  sList: TStringList;
+begin
+  sList := TStringList.Create;
+  sList.Text:=sHeader;
+  sList.LineBreak:=#10;
+  sList.NameValueSeparator:=':';
+  sList.Sorted:=True;
+  sList.Sort;
+  Result := '';
+  for i := 1 to sList.Count - 1 do
+    Result := Result + sList[i]+#10;
+
+end;
+
+procedure TAWSSignatureVersion4.SignedHeaders(const sHeader: String; var sToSing, sToCanonical: String);
+var
+  i: Integer;
+  sList: TStringList;
+  sName, sValue: String;
+begin
+  sList := TStringList.Create;
+  sList.Text:=sHeader;
+  sList.LineBreak:=#10;
+  sList.NameValueSeparator:=':';
+  sList.Sorted:=True;
+  sList.Sort;
+  for i := 1 to sList.Count - 1 do
+    begin
+      sList.GetNameValue(i, sName, sValue);
+      sToSing := sToSing + LowerCase(sName) + ';';
+      sToCanonical := sToCanonical + LowerCase(sName)+':'+sValue+#10;
+    end;
+  system.Delete(sToSing, Length(sToSing), 1);
+end;
+
+function TAWSSignatureVersion4.Calculate(Request: IHTTPRequest): string;
```

Comment:
@almp1 this method is complex. Is it possible to refactor, creating some other private methods?